pkg/handler/metricsconfigurator/handler.go

```diff
@@ -0,0 +1,342 @@
+	}
+	logger.V(4).Info("parsed response, metric filters", "filters", MetricsFilters)
+
+	// Create a cloudwatch client
+	cwClient := cloudwatch.NewFromConfig(cfg)
+
+	// add filters to provided metric stream
+	_, err = cwClient.PutMetricStream(ctx, &cloudwatch.PutMetricStreamInput{
+		FirehoseArn:    &h.FirehoseArn,
+		RoleArn:        &h.RoleArn,
+		OutputFormat:   types.MetricStreamOutputFormat(h.OutputFormat),
+		Name:           &h.MetricStreamName,
+		IncludeFilters: MetricsFilters,
+	})
+	if err != nil {
+		return nil, h.reportAndError("failed to add filters to metric stream", req, err)
+	}
+	logger.V(4).Info("successfully added filters to metric stream")
+
+	err = h.reportStatus(*req, true, "successfully wrote metrics to metric stream")
+	if err != nil {
+		return nil, fmt.Errorf("failed to report status to cloudformation: %w, during successful write", err)
+	}
+
+	logger.V(3).Info("returned response to cloudformation")
+	return []byte{}, nil
+}
+
+func (h Handler) parsePayload(payload []byte) (*Request, error) {
+
+	var req Request
+	// we cannot report errors until we have parsed out the json for the request,
+	// which contains the url to send the response to
+	if err := json.Unmarshal(payload, &req); err != nil {
+		return nil, fmt.Errorf("failed to decode payload for metricsconfigurator: %w", err)
+	}
+	return &req, nil
+}
+
+func (h Handler) parseResponse(bodyBytes []byte) ([]types.MetricStreamFilter, error) {
+	logger := h.Logger
+	var result GraphQLResponse
+	err := json.Unmarshal(bodyBytes, &result)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling JSON from GQL response: %w", err)
+	}
+
+	logger.V(4).Info("response from observe api", "result", result)
+
+	var metricSelection []MetricsListItem
+
+	variablesList := result.Data.Datasource.Variables
+
+	var targetVariable *Variable
+
+	// Iterate through the variables and find one with the name "Metrics"
+	// This is where we should find the metrics configuration
+	for _, variable := range variablesList {
+		if variable.Name == "Metrics" {
+			targetVariable = &variable
+			break
+		}
+	}
+
+	if targetVariable == nil {
+		return nil, fmt.Errorf("metrics variable not set in datasource, %+v", result)
+	}
+
+	metricSelection = targetVariable.Details.AwsCollectionStackConfig.AwsServiceMetricsList
+
+	MetricsFilters := convertToMetricStreamFilters(metricSelection)
+	return MetricsFilters, nil
+}
+
+func convertToMetricStreamFilters(MetricsList []MetricsListItem) []types.MetricStreamFilter {
+
+	// to ensure that an empty slice is returned instead of nil when `MetricsList` is empty
+	metricsFilters := make([]types.MetricStreamFilter, 0)
+	for _, service := range MetricsList {
+		metricsFilters = append(metricsFilters, types.MetricStreamFilter{
+			Namespace:   &service.Namespace,
+			MetricNames: service.MetricNames,
+		})
+	}
+
+	return metricsFilters
+}
+
+func (h Handler) getSecretValue(ctx context.Context, cfg aws.Config) (*string, error) {
+
```

Comment:
nit: there seems to be a lot of whitespace right after the function definition line? rm